design/hotshard.md

```diff
@@ -0,0 +1,385 @@
+  - [Avoiding Read Hotspots](#avoiding-read-hotspots)
+  - [Schema & Keyspace Design Patterns](#schema--keyspace-design-patterns)
+  - [What to do when encountered in production](#what-to-do-when-encountered-in-production)
+  - [Optional Hot Shard Throttling (Targeted Mitigation)](#optional-hot-shard-throttling-targeted-mitigation)
+  - [Key Insight](#key-insight)
+- [Server Knobs Related to Hot Shards](#server-knobs-related-to-hot-shards)
+  - [Hot Shard Throttling Knobs](#hot-shard-throttling-knobs)
+  - [Shard Bandwidth Splitting Knobs](#shard-bandwidth-splitting-knobs)
+  - [Read Hot Shard Detection Knobs](#read-hot-shard-detection-knobs)
+  - [Storage Queue Rebalancing Knobs](#storage-queue-rebalancing-knobs)
+  - [Read Rebalancing Knobs](#read-rebalancing-knobs)
+  - [Shard Split/Merge Priority Knobs](#shard-splitmerge-priority-knobs)
+  - [Other Hot Shard Related Knobs](#other-hot-shard-related-knobs)
+  - [Key Relationships](#key-relationships)
+- [References](#references)
+
+## Overview
+
+FoundationDB partitions the keyspace into shards (contiguous key ranges) and assigns each shard to a storage-team replica set.
+
+Shards are just key ranges. Their size adapts continuously as Data Distribution monitors load and splits or merges ranges to stay within configured bounds [1][2]. Splits are triggered when metrics like bytes, read/write bandwidth, or hot-shard detection exceed thresholds (splitMetrics.bytes = shardBounds.max.bytes / 2, splitMetrics.bytesWrittenPerKSecond etc.). In other words, shard size is workload driven rather than fixed [1][3]. Storage servers and commit proxies provide ongoing measurements so DD can resize ranges dynamically to relieve load hotspots and maintain balance [4][5].
+
+The maximum shard size is:
+```
+int64_t getMaxShardSize(double dbSizeEstimate) {
+    int64_t size = std::min((SERVER_KNOBS->MIN_SHARD_BYTES + (int64_t)std::sqrt(std::max<double>(dbSizeEstimate, 0)) *
+      SERVER_KNOBS->SHARD_BYTES_PER_SQRT_BYTES) *
+      SERVER_KNOBS->SHARD_BYTES_RATIO,
+        (int64_t)SERVER_KNOBS->MAX_SHARD_BYTES);
+```
+  | dbSizeEstimate | Calculation                                | Result (approx.) |
+  |----------------|--------------------------------------------|------------------|
+  | 0              | (10 MB + 0) × 4                            | 40 MB            |
+  | 1 GB           | (10 MB + √1 GB × 45) × 4                   | 45.7 MB          |
+  | 100 GB         | (10 MB + √100 GB × 45) × 4                 | 96.9 MB          |
+  | 1 TB           | (10 MB + √1 TB × 45) × 4                   | 220 MB           |
+  | ≈6.5 TB        | (10 MB + √6.5 TB × 45) × 4 → hits cap      | 500 MB           |
+  | ≥10 TB         | Any larger value → capped by MAX_SHARD_BYTES | 500 MB         |
+
+## What is a Hot Shard?
+
+A shard is "hot" when it absorbs a disproportionate share of the cluster's write workload (not reads!), driving CPU or bandwidth saturation on its replica servers [6].
```

Comment:
Why is it "impossible" to have a read hot shard? The referenced doc at least speaks of write-hot shards and read-hot shards: https://github.com/apple/foundationdb/blob/7.3.0/design/data-distributor-internals.md#which-shard-to-move. 

A few lines below this doc talks about read-hot shards: `... thresholds is tagged as read-hot so the data distributor process ...`.

Reply:
I will clarify this to say that we only split on write hot shards. We do not split on read hot shards. Read-hot detection impacts the relocation logic rather than splitting.